api/lightwallet_rest.md

```diff
@@ -400,3 +445,95 @@ Submit raw transaction to be relayed to monero network.
+| maj_i *   | `uint32`         | Subaddress major index (defaults to 0)          |
+| min_i *   | `uint32`         | Subaddress minor index (defaults to 0)          |
+| n_maj *   | `uint32`         | Number of major subaddresses to provision       |
+| n_min *   | `uint32`         | Number of minor subaddresses to provision       |
+| get_all * | `boolean`        | Whether to include all subaddresses in response |
+
+> The various combinations of `maj_i`, `min_i`, `n_maj`, and `n_min` behave
+> differently, but generally the server provisions subaddresses where it has not
+> already provisioned them, with the indexes specified as lower bounds.
+
+> If, for example, only `maj_i` is included, then the server should provision a
+> default number of minor subaddresses (e.g. 500) within that `maj_i`, wherever
+> minor subaddresses have not already been provisioned, starting with `min_i`
+> set to 0 as the lower bound.
+
+> If, for example, `maj_i`, `min_i`, `n_maj` and `n_min` are included, then the
+> server should provision `n_maj` majors wherever major subaddresses have not
+> already been provisioned starting with `maj_i` as the lower bound, and for
+> each major, provision `n_min` subaddresses starting with `min_i` as the lower
+> bound.
+
+> All combinations follow the above framework.
+
+> The server can choose to "pad" counts and provision *more* than a client
+> requests.
+
+> If the server cannot provision a specified number of subaddresses because the
+> server would provision more than the maximum number of subaddresses, HTTP 409
+> should be returned. Ack that the status code is not perfect here. In this
+> case, the client should make a new request either with a smaller requested
+> number of subaddresses to provision, or at a different major or minor index.
+
+> If none of `maj_i`, `min_i`, `n_maj`, and `n_min` are included in the request,
+> the server must return a HTTP 400 "Bad Request" error.
+
+> `get_all` defaults to true if not included in the request.
+
+**Response** object
+
+|     Field      |    Type    |                          Description                        |
+|----------------|------------|-------------------------------------------------------------|
+| new_subaddrs   | `subaddrs` | All new subaddresses provisioned in the request             |
+| all_subaddrs * | `subaddrs` | All subaddresses provisioned for the wallet (including new) |
+
+#### `upsert_subaddrs`
+Upsert subaddresses at the specified major and minor indexes. This endpoint is
+idempotent.
+
+**Request** object
+
+|   Field   |       Type       |                  Description                    |
+|-----------|------------------|-------------------------------------------------|
+| address   | `base58-address` | Standard address of the wallet                  |
+| view_key  | `binary`         | View key bytes                                  |
+| subaddrs  | `subaddrs`       | Subaddresses to upsert                          |
```

Comment:
What happens if it already exists? I suppose this is the difference between the other API call?

Reply:
Upsert as in update unless exists in which case insert. Server should return its known set of subaddrs so client can know, for various reasons, which I will describe soon in some updates.

Over the past 1-2 years, I have done a massive amount of work on the LWS and discovered that the subaddr support in the LWS API as currently specced and as it even currently behaves is badly insufficient, so I will share a list of things I've discovered with some spec proposals shortly. I'm finishing it up now after collecting everything.

Reply:
insert unless exists*

Reply:
@j-berman is this a single `subaddrs` or an array? I'm assuming an array because the other instances (`new_subaddrs` and `all_subaddrs`) have to be arrays by design.

Reply:
@vtnerd it needs to be what i call a complex range of idxs. that is, a dictionary with maj idx str keys pointing to arrays of arrays of min idxs.

Reply:
That sounds nearly identical to this design. A major index, pointing to an array of index ranges.

Reply:
well, it was your question, not anything major I'm claiming to point out yet, but btw I did intend to type array of arrays. you need support for discontiguous min idx ranges.

Reply:
This API does support discontiguous min ranges. `[10, [0,2],[4,7]]` means `0,1,2,4,5,6,7` minor indexes with major index `10`.

Reply:
but that is a discontiguous min range.